## Creation of the dataset

We have two datasets
 - Amazon Customer Review dataset :
   - amazon_reviews_us_Books_v1_00.tsv.gz (2.74 Go)
   - amazon_reviews_us_Books_v1_01.tsv.gz (2.69 Go)
   - amazon_reviews_us_Books_v1_02.tsv.gz (1.33 Go)
 - Amazon Book Cover dataset : 
   - book32-listing.csv (39.5 Mo)

Each row of the first dataset represents a review. Therefore we could have several rows representing the same book (several reviews for the same book). 

Each row of the second dataset represents a book. 

In this notebook, we merge the two datasets, in order to have each row of the new dataset representing a review, along with the cover of the corresponding book. 

Since the first dataset is pretty big, and to avoid creating a huge dataset, we will use the following pipeline : 
 - Load the cover file
 - Load the first review file
 - Merge the first review file in df_merge_1
 - Load the second review file
 - Merge the second review file in df_merge_2
 - Load the third review file
 - Merge the third review file in df_merge_3
 - Merge df_merge_1, df_merge_2, and df_merge_3 in df_merge
 - Save df_merge in merge.csv

In [5]:
import pandas as pd
import numpy as np

First we load the cover image dataset

In [6]:
columns = ['product_id', 
           'filename', 
           'image_url', 
           'title', 
           'author', 
           'category_id', 
           'category']
df_covers = pd.read_csv("Data/book32-listing.csv", 
                        encoding = "ISO-8859-1", 
                        header=None, 
                        names=columns)
df_covers.head()

product_id        filename  \
0   761183272  0761183272.jpg   
1  1623439671  1623439671.jpg   
2  B00O80WC6I  B00O80WC6I.jpg   
3   761182187  0761182187.jpg   
4  1578052084  1578052084.jpg   

                                           image_url  \
0  http://ecx.images-amazon.com/images/I/61Y5cOdH...   
1  http://ecx.images-amazon.com/images/I/61t-hrSw...   
2  http://ecx.images-amazon.com/images/I/41X-KQqs...   
3  http://ecx.images-amazon.com/images/I/61j-4gxJ...   
4  http://ecx.images-amazon.com/images/I/51Ry4Tsq...   

                                               title              author  \
0                    Mom's Family Wall Calendar 2016      Sandra Boynton   
1                    Doug the Pug 2016 Wall Calendar        Doug the Pug   
2  Moleskine 2016 Weekly Notebook, 12M, Large, Bl...           Moleskine   
3            365 Cats Color Page-A-Day Calendar 2016  Workman Publishing   
4               Sierra Club Engagement Calendar 2016         Sierra Club   

   category_id   category  
0            3  Calendars  
1            3  Calendars  
2            3  Calendars  
3            3  Calendars  
4            3  Calendars

We read the dataset containing the reviews. We have to specify the columns names because some rows have a higher count of cell than others. We want to see what is contained in these rows. 

In [7]:
cols = ["marketplace",
        "customer_id",
        "review_id",
        "product_id",
        "product_parent",
        "product_title",
        "product_category",
        "star_rating",
        "helpful_votes",
        "total_votes",
        "vine",
        "verified_purchase",
        "review_headline",
        "review_body",
        "review_date", 
        "undefined_1",
        "undefined_2",
        "undefined_3",
        "undefined_4",
        "undefined_5",
        "undefined_6",
        "undefined_7"
       ]

In [8]:
df_reviews = pd.read_csv("Data/amazon_reviews_us_Books_v1_00.tsv.gz", 
                         names=cols, 
                         sep='\t', 
                         skiprows=1)

/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (7,17,18,19,20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


The warning simply indicates that some columns do not have the same dtype for each row, this will be handled later by cleaning the data.

So first we check the rows with a higher count of cells.

In [9]:
df_reviews[df_reviews["undefined_1"].notnull()]

marketplace  customer_id       review_id  product_id  product_parent  \
3522             US     45223684  R1NHKVD5LMP2JY  1503551814       353548609   
5280             US     18701079  R1946J1OJKC1K2  0743292472         7685359   
20476            US     40945620  R33MH6R1TZWIGR  1452572747       523136939   
25893            US     22921197  R1OPQAX1D7DWB8  0984755101       632337521   
27014            US     44087253  R24H2K78E883A0  1616496169       709191612   
59796            US     44250036   RYNEKW5P1Z3AG  0385262078       887509044   
69196            US     50641462  R2VI5KEAQR2P23  0545436826       247155770   
71951            US     43676679  R2F0SUDYHR79JQ  1844094642       766424526   
78718            US      4881222  R21NUSCF1WC1I8  1607512572       236705009   
81298            US      1148395  R1UH3R71GX97AR  1502716984       563128335   
87681            US     50861690  R2Z0YTRQKVFHSF  0807403083       147336042   
90514            US     28736875   RVFNQHOLE58A0  0806120940       661253970   
91145            US     15629461    R4U6722KCUT6  1592408419       543673826   
108943           US      8612032  R2V2ES66TGP7RN  1460268601       273651829   
118287           US     50953132  R2YWTZB0TCB920  1439132852       537451993   
124665           US     38871814  R1CQPRDZKZZKKN  097860783X       764135422   
125539           US     26965291  R33FTU1VADZM18  1591843820       618286357   
144113           US     19357930  R370TRRWYXD4HH  0743230418       757722318   
167694           US     49384329  R2ZVVLUZWJ1EWP  1502328585       460342874   
179854           US     51750629  R1919SOCLXFHH3  1456550144       598402352   
189710           US     41990563   RBWE5V226JS4A  0762454865       402819053   
196486           US     20804529  R1W0RQ0ETQT7V9  1475812140       387091568   
232130           US     26084626   RX8T8JKA1RCM4  0062333933       223598514   
241732           US     36457256  R32B42EMWA43S7  1400076560       204550997   
269921           US     17688934   R70Y8Z54OL2OW  0316210463       723771425   
279156           US     12975716  R3KDGN68ECWPFL  1285755618       475852195   
283355           US     15648568  R22WHMC556RKAR  386521584X       467228901   
285765           US     11380786  R12VP0F0Q16NMF  0971339309       192326528   
308164           US     41713493  R1EW4GKVIF0ALE  1503019500       768980686   
315197           US     45994291   RK3U2RRENGDFM  1511510056       439740394   
...             ...          ...             ...         ...             ...   
4449651          US      2484994  R3K2FPQZ7X8UVP  1497437016       519821647   
4459300          US     40203782  R1SILLK0ZNLGSD  1434238822       755154206   
4477211          US     14426796  R3HBOV96BGZSP2  0870717421       769808046   
4500861          US     13558114  R2F60Z8M7MYER2  038573252X       305985379   
4549884          US     46366696  R2PAOB044HM77A  1492927449       113630502   
4552286          US      4274094   RXS1XC8LMKG1Y  0823422437       304889763   
4558857          US     27197209  R2GNMOUPSY80XK  1611861314       616413203   
4562249          US     49569659   R7N4GP96VKLSA  9231039911       306686978   
4567912          US     52956445  R13129ZA040001  0470376295       280360616   
4633359          US     37024654   RGSFJRNLHB30C  0199654441       755550981   
4640815          US     19973002  R2OQPGHTYWVCZ2  0991067886       313246709   
4644915          US     30817773  R1WAHI6773B6KO  1497511135       946709255   
4671851          US     32651993  R3UV6QM6ZDKGYU  0385184581       212671778   
4701152          US      6071575   R9V19CKJTF8GG  055380653X       483985572   
4721068          US     38970700   RRGRPELJQGFI6  0822210738       994387679   
4731354          US     36857681  R3S1ZPUV914YNM  0761536434       630858517   
4744853          US     44416620  R1ZCFR2H9H98QN  1402720602        84227631   
4939718          US     47464628  R1RTM2W1I88ZGU  031074265X       972127661   
5184881 

In [10]:
# Compute the proportion of long row
null_mask = df_reviews["undefined_1"].notnull()
print(null_mask.sum() )
print((~null_mask).sum())

326
10236850


What is contained in the undefined columns is a bit weird, it's like we have more reviews, but they are not really related to the standard review column. For example, in the second line of the output, the review is "Was a great read..." and then we have "NOT PLEASED WITH THE VENDORS AT ALL" etc.

These lines are not clear concerning the reviews. We also try to find these reviews on the amazon website. We found some of them, but we were only able to find the main reviews, and not the reviews contained in the additional columns. Besides, these rows are not coherent concerning the columns (some values represents something else than expected by the column name). Since we only have 326 of such rows over 10 millions, we decide to drop these rows. We can also drop the additional column since we won't need them anymore. 

Finally, we drop some other columns that we don't need 

In [11]:
df_reviews = df_reviews[~null_mask]
df_reviews = df_reviews.iloc[:, :15]
df_reviews = df_reviews.drop(['product_parent', 
                              'product_category', 
                              'vine', 
                              'verified_purchase'], axis=1)

In [12]:
df_reviews["product_id"] = df_reviews["product_id"].astype('str')

Now we want to merge the reviews dataset with the covers dataset, in other word, for each reviews, we want to have the cover image of the book. 

In [13]:
df_merge_1 = df_reviews.merge(df_covers, 
                              left_on='product_id', 
                              right_on='product_id', 
                              how='inner')

In [14]:
print(len(df_merge_1))

1409974


From the 10M reviews, we have 1.4M that corresponds to a book from the 137'000 books of the covers dataset. The loss is consequent, but we still have a important dataset to work on. Moreover, we still have the two others reviews dataset to include in this dataframe.

Note that we prefered not to load all three reviews dataset because of memory concerns. Our pipeline is : load covers, load review_1, merge, save merge as csv, load review_2, merge, save merge as csv, load review_3, merge, save merge as csv. 

In [15]:
df_reviews = pd.read_csv("Data/amazon_reviews_us_Books_v1_01.tsv.gz", 
                         names=cols, 
                         sep='\t', 
                         skiprows=1)

/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (7,17,18,19,20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [16]:
null_mask = df_reviews["undefined_1"].notnull()
print(null_mask.sum() )
print((~null_mask).sum())

1
6106190


In [17]:
df_reviews = df_reviews[~null_mask]
df_reviews = df_reviews.iloc[:, :15]
df_reviews = df_reviews.drop(['product_parent', 
                              'product_category', 
                              'vine', 
                              'verified_purchase'], axis=1)

In [18]:
df_reviews["product_id"] = df_reviews["product_id"].astype('str')

In [19]:
df_merge_2 = df_reviews.merge(df_covers, 
                              left_on='product_id', 
                              right_on='product_id', 
                              how='inner')

In [20]:
print(len(df_merge_2))

359693


Finally the third reviews dataset

In [21]:
df_reviews = pd.read_csv("Data/amazon_reviews_us_Books_v1_02.tsv.gz", 
                         names=cols, 
                         sep='\t', 
                         skiprows=1)

/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (17,18,19,20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [22]:
null_mask = df_reviews["undefined_1"].notnull()
print(null_mask.sum() )
print((~null_mask).sum())

1
3105370


In [23]:
df_reviews = df_reviews[~null_mask]
df_reviews = df_reviews.iloc[:, :15]
df_reviews = df_reviews.drop(['product_parent', 
                              'product_category', 
                              'vine', 
                              'verified_purchase'], axis=1)

In [24]:
df_reviews["product_id"] = df_reviews["product_id"].astype('str')

In [25]:
df_merge_3 = df_reviews.merge(df_covers, 
                            left_on='product_id', 
                            right_on='product_id', 
                            how='inner')

In [26]:
print(len(df_merge_3))

72122


In [27]:
# append the dataframes
df_merge = df_merge_1.append(df_merge_2).append(df_merge_3)

In [28]:
df_merge.to_csv("Data/merge.csv", sep='\t', index=False)